### PR TITLE
feat: allow free input in AutoComplete

### PR DESCRIPTION
**Pull request by qrilka**
Hi @luoxiaozero here is my monkey-patched version of AutoComplete - I needed to support free input and not only selection from `options`.
Maybe this could be made into an option of `AutoComplete`, what do you think?

### CODE REVIEW

**Comment by qrilka** [comment]
@luoxiaozero any comments? I've also added "allow_free_input" as an option. So basically by default it almost work as a previous version. Probably the main difference is that the first option doesn't get selected by default and also options update clears up selection (the current component behaves confusingly in this respect in my opinion)

**Comment by luoxiaozero** [comment]
I think when allow_free_input is false, the previous behavior can be maintained (as long as the option is not empty, then select_option_index cannot be None)

**Comment by luoxiaozero** [comment]
When the option changes, allow_free_input is 1 if it is false and None if it is true.

**Comment by qrilka** [comment]
I've also switched the option to just a boolean, not a signal. The source motivation was compile error about NotRawCallback but in principle it makes very little sense to make this option reactive.

File: thaw/src/auto_complete/mod.rs [hunk 1]
```diff
@@ -82,39 +85,59 @@ pub fn AutoComplete(
...
-                    *index = options.with_untracked(|options| options.len()) - 1;
+                if *index == Some(0) {
+                    if allow_free_input {
+                        *index = None
+                    } else {
+                        *index = Some(options.with_untracked(|options| options.len()) - 1);
+                    }
                 } else {
-                    *index -= 1
+                    *index = index.map(|index| index - 1)
```
Comment (luoxiaozero) [inline thaw/src/auto_complete/mod.rs:124]: There's a marginal problem. When index is None, the last of the array should be selected.

Comment (qrilka) [reply thaw/src/auto_complete/mod.rs:124]: I'm not quite sure if that would be intuitive but I'll switch to that.

File: thaw/src/auto_complete/mod.rs [hunk 2]
```diff
@@ -82,39 +85,59 @@ pub fn AutoComplete(
...
+            select_option_index.set(None);
+        }
         is_show_menu.set(false);
     };
 
+    // we unset selection index whenever options get changed
+    // otherwise e.g. selection could move from one item to
+    // another staying on the same index
+    create_effect(move |_| {
+        options.get();
```
[the remark this answers is not in the document]
Comment (qrilka) [reply thaw/src/auto_complete/mod.rs:98]: Thanks, I'm still new to leptos...

File: thaw/src/auto_complete/mod.rs (hunk 1 above)
Comment (luoxiaozero) [reply thaw/src/auto_complete/mod.rs]: Github's search box works the same way.

**Comment by luoxiaozero** [comment]
Thanks!

**Pull request merged**